core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala

```diff
@@ -2177,6 +2177,46 @@ class BlockManagerSuite extends SparkFunSuite with Matchers with BeforeAndAfterE
     assert(kryoException.getMessage === "java.io.IOException: Input/output error")
   }
 
+  test("SPARK-39647: Failure to register with ESS should prevent registering the BM") {
+    val timeoutExec = "timeoutExec"
+    val handler = new NoOpRpcHandler {
+      override def receive(
+          client: TransportClient,
+          message: ByteBuffer,
+          callback: RpcResponseCallback): Unit = {
+        val msgObj = BlockTransferMessage.Decoder.fromByteBuffer(message)
+        msgObj match {
+
+          case exec: RegisterExecutor if exec.execId == timeoutExec =>
+            () // No reply to generate client-side timeout
+        }
+      }
+    }
+    val transConf = SparkTransportConf.fromSparkConf(conf, "shuffle", numUsableCores = 0)
+    Utils.tryWithResource(new TransportContext(transConf, handler, true)) { transCtx =>
+      // a server which delays response 50ms and must try twice for success.
```

Comment:
This comment is copy-pasted from "SPARK-20640: Shuffle registration timeout and maxAttempts conf are working" but here it is not a valid statement.